File: sdks/community/ruby/test/ag_ui_protocol/core/types_test.rb
require "test_helper"
require "json"

class TypesTest < Minitest::Test
  context "AgUiProtocol::Core::Types" do
    context "Model" do
      should "don't serialize to JSON because it's abstract" do
        assert_raises(NotImplementedError) do
          AgUiProtocol::Core::Types::Model.new.to_json
        end
      end

      should "raise when unknown keyword is provided" do
        assert_raises(ArgumentError) do
          AgUiProtocol::Core::Types::Model.new(unknown: 1)
        end
      end
    end

    context "FunctionCall" do
      should "serialize to JSON" do
        obj = AgUiProtocol::Core::Types::FunctionCall.new(name: "f", arguments: "{}")
        payload = JSON.parse(obj.to_json)
        assert_equal "f", payload["name"]
      end

      should "raise when name is missing" do
        assert_raises(ArgumentError) do
          AgUiProtocol::Core::Types::FunctionCall.new(arguments: "{}")
        end
      end
    end

    context "ToolCall" do
      should "serialize to JSON" do
        obj = AgUiProtocol::Core::Types::ToolCall.new(id: "tc1", function: { name: "f", arguments: "{}" })
        payload = JSON.parse(obj.to_json)
        assert_equal "tc1", payload["id"]
        assert_equal "function", payload["type"]
      end

      should "raise when function is missing" do
        assert_raises(ArgumentError) do
          AgUiProtocol::Core::Types::ToolCall.new(id: "tc1")
        end
      end
    end

    context "BaseMessage" do
      should "serialize to JSON" do
        obj = AgUiProtocol::Core::Types::BaseMessage.new(id: "m1", role: "assistant", content: "hi")
        payload = JSON.parse(obj.to_json)
        assert_equal "assistant", payload["role"]
      end

      should "raise when role is missing" do
        assert_raises(ArgumentError) do
          AgUiProtocol::Core::Types::BaseMessage.new(id: "m1")
        end
      end
    end

    context "DeveloperMessage" do
      should "serialize to JSON" do
        obj = AgUiProtocol::Core::Types::DeveloperMessage.new(id: "d1", content: "hi")
        payload = JSON.parse(obj.to_json)
        assert_equal "developer", payload["role"]
      end

      should "raise when content is missing" do
        assert_raises(ArgumentError) do
          AgUiProtocol::Core::Types::DeveloperMessage.new(id: "d1")
        end
      end
    end

    context "SystemMessage" do
      should "serialize to JSON" do
        obj = AgUiProtocol::Core::Types::SystemMessage.new(id: "s1", content: "hi")
        payload = JSON.parse(obj.to_json)
        assert_equal "system", payload["role"]
      end

      should "raise when content is missing" do
        assert_raises(ArgumentError) do
          AgUiProtocol::Core::Types::SystemMessage.new(id: "s1")
        end
      end
    end

    context "AssistantMessage" do
      should "serialize to JSON" do
        obj = AgUiProtocol::Core::Types::AssistantMessage.new(
          id: "a1",
          content: "hi",
          tool_calls: [{ id: "tc1", function: { name: "f", arguments: "{}" } }]
        )
        payload = JSON.parse(obj.to_json)
        assert_equal "assistant", payload["role"]
      end

      should "raise when tool_calls are invalid" do
        assert_raises(ArgumentError) do
          AgUiProtocol::Core::Types::AssistantMessage.new(id: "a1", tool_calls: [{ id: "tc1" }])
        end
      end
    end

    context "TextInputContent" do
      should "serialize to JSON" do
        obj = AgUiProtocol::Core::Types::TextInputContent.new(text: "hello")
        payload = JSON.parse(obj.to_json)
        assert_equal "text", payload["type"]
      end

      should "raise when text is missing" do
        assert_raises(ArgumentError) do
          AgUiProtocol::Core::Types::TextInputContent.new
        end
      end
    end

    context "BinaryInputContent" do
      should "raise when no source is provided" do
        assert_raises(ArgumentError) do
          AgUiProtocol::Core::Types::BinaryInputContent.new(mime_type: "image/png")
        end
      end

      should "serialize with camelCase" do
        content = AgUiProtocol::Core::Types::BinaryInputContent.new(mime_type: "image/png", url: "https://example.com/a.png")
        payload = JSON.parse(content.to_json)

        assert_equal "binary", payload["type"]
        assert_equal "image/png", payload["mimeType"]
        assert_equal "https://example.com/a.png", payload["url"]
        refute payload.key?("data")
      end

      should "raise when mime_type is missing" do
        assert_raises(ArgumentError) do
          AgUiProtocol::Core::Types::BinaryInputContent.new(url: "https://example.com/a.png")
        end
      end
    end

    context "UserMessage" do
      should "serialize to JSON" do
        msg = AgUiProtocol::Core::Types::UserMessage.new(id: "u1", content: "hello")
        payload = JSON.parse(msg.to_json)
        assert_equal "user", payload["role"]
      end

      should "raise when id is missing" do
        assert_raises(ArgumentError) do
          AgUiProtocol::Core::Types::UserMessage.new(content: "hello")
        end
      end

      should "normalize array content into typed input content models" do
        msg = AgUiProtocol::Core::Types::UserMessage.new(
          id: "u1",
          content: [
            { type: "text", text: "hello" },
            { type: "binary", mimeType: "image/png", url: "https://example.com/a.png" }
          ]
        )

        assert_kind_of Array, msg.content
        assert_kind_of AgUiProtocol::Core::Types::TextInputContent, msg.content[0]
        assert_kind_of AgUiProtocol::Core::Types::BinaryInputContent, msg.content[1]
      end

      should "serialize content with camelCase keys" do
        msg = AgUiProtocol::Core::Types::UserMessage.new(
          id: "u1",
          content: [
            { type: "binary", mimeType: "image/png", url: "https://example.com/a.png", filename: nil }
          ]
        )

        payload = JSON.parse(msg.to_json)

        assert_equal "u1", payload["id"]
        assert_equal "user", payload["role"]
        assert_kind_of Array, payload["content"]
        assert_equal "binary", payload["content"][0]["type"]
        assert_equal "image/png", payload["content"][0]["mimeType"]
        assert_equal "https://example.com/a.png", payload["content"][0]["url"]
        refute payload["content"][0].key?("filename")
      end
    end

    context "ToolMessage" do
      should "serialize to JSON" do
        msg = AgUiProtocol::Core::Types::ToolMessage.new(id: "tm1", content: "ok", tool_call_id: "tc1")
        payload = JSON.parse(msg.to_json)
        assert_equal "tool", payload["role"]
      end

      should "raise when tool_call_id is missing" do
        assert_raises(ArgumentError) do
          AgUiProtocol::Core::Types::ToolMessage.new(id: "tm1", content: "ok")
        end
      end
    end

    context "ActivityMessage" do
      should "serialize to JSON" do
        msg = AgUiProtocol::Core::Types::ActivityMessage.new(id: "am1", activity_type: "progress", content: { "pct" => 10 })
        payload = JSON.parse(msg.to_json)
        assert_equal "activity", payload["role"]
      end

      should "raise when activity_type is missing" do
        assert_raises(ArgumentError) do
          AgUiProtocol::Core::Types::ActivityMessage.new(id: "am1", content: { "pct" => 10 })
        end
      end
    end

    context "Context" do
      should "serialize to JSON" do
        obj = AgUiProtocol::Core::Types::Context.new(description: "d", value: "v")
        payload = JSON.parse(obj.to_json)
        assert_equal "d", payload["description"]
      end

      should "raise when value is missing" do
        assert_raises(ArgumentError) do
          AgUiProtocol::Core::Types::Context.new(description: "d")
        end
      end
    end

    context "Tool" do
      should "serialize to JSON" do
        obj = AgUiProtocol::Core::Types::Tool.new(name: "t", description: "d", parameters: { "type" => "object" })
        payload = JSON.parse(obj.to_json)
        assert_equal "t", payload["name"]
      end

      should "raise when parameters is missing" do
        assert_raises(ArgumentError) do
          AgUiProtocol::Core::Types::Tool.new(name: "t", description: "d")
        end
      end
    end

    context "RunAgentInput" do
      should "serialize to JSON" do
        obj = AgUiProtocol::Core::Types::RunAgentInput.new(
          thread_id: "t1",
          run_id: "r1",
          state: {},
          messages: [],
          tools: [],
          context: [],
          forwarded_props: {}
        )
        payload = JSON.parse(obj.to_json)
        assert_equal "t1", payload["threadId"]
      end

      should "raise when forwarded_props is missing" do
        assert_raises(ArgumentError) do
          AgUiProtocol::Core::Types::RunAgentInput.new(thread_id: "t1", run_id: "r1", state: {}, messages: [], tools: [], context: [])
        end
      end
    end

    should "work with array content" do
      messages = [
        AgUiProtocol::Core::Types::UserMessage.new(id: "u1", content: "hello"),
        AgUiProtocol::Core::Types::AssistantMessage.new(id: "a1", content: "hi"),
        AgUiProtocol::Core::Types::ActivityMessage.new(id: "am1", activity_type: "progress", content: { "pct" => 10 }),
      ]
      tools = [
        AgUiProtocol::Core::Types::Tool.new(name: "t1", description: "d1", parameters: { "type" => "object" }),
      ]
      context = [
        AgUiProtocol::Core::Types::Context.new(description: "d1", value: "v1"),
      ]
      obj = AgUiProtocol::Core::Types::RunAgentInput.new(
        thread_id: "t1",
        run_id: "r1",
        state: {},
        messages: messages,
        tools: tools,
        context: context,
        forwarded_props: {}
      )
      payload = JSON.parse(obj.to_json)
      assert_equal "t1", payload["threadId"]
      assert_equal "r1", payload["runId"]
      assert_equal "u1", payload["messages"][0]["id"]
      assert_equal "a1", payload["messages"][1]["id"]
      assert_equal "am1", payload["messages"][2]["id"]
    end

    should "raise when messages type is invalid" do
      messages = [
        AgUiProtocol::Core::Types::TextInputContent.new(text: "hello"),
        AgUiProtocol::Core::Types::FunctionCall.new(name: "f", arguments: "{}"),
        AgUiProtocol::Core::Types::ToolCall.new(id: "tc1", function: { name: "f", arguments: "{}" }),
      ]
      
      assert_raises(ArgumentError) do
        obj = AgUiProtocol::Core::Types::RunAgentInput.new(
        thread_id: "t1",
        run_id: "r1",
        state: {},
        messages: messages,
        tools: [],
        context: [],
        forwarded_props: {}
      )
      end
    end
  end
end

File: sdks/community/go/pkg/core/types/types_test.go
package types

import (
	"encoding/json"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// TestRunAgentInputUnmarshalCamelCase verifies decoding a camelCase RunAgentInput payload.
func TestRunAgentInputUnmarshalCamelCase(t *testing.T) {
	payload := []byte(`{
		"threadId": "thread-1",
		"runId": "run-1",
		"parentRunId": "run-0",
		"state": {"mode": "test"},
		"messages": [
			{"id": "msg-1", "role": "user", "content": "hello"},
			{
				"id": "msg-2",
				"role": "assistant",
				"content": "hi",
				"encryptedContent": "enc-content-msg-2",
				"toolCalls": [
					{
						"id": "tc-1",
						"type": "function",
						"function": {"name": "tool", "arguments": "{}"}
					}
				]
			},
			{
				"id": "reasoning-1",
				"role": "reasoning",
				"content": "summary",
				"encryptedValue": "enc-reasoning-1"
			}
		],
		"tools": [{"name": "tool", "description": "desc", "parameters": {"type": "object"}}],
		"context": [{"description": "ctx", "value": "val"}],
		"forwardedProps": {"traceId": "abc"}
	}`)

	var input RunAgentInput
	err := json.Unmarshal(payload, &input)
	require.NoError(t, err)

	assert.Equal(t, "thread-1", input.ThreadID)
	assert.Equal(t, "run-1", input.RunID)
	require.NotNil(t, input.ParentRunID)
	assert.Equal(t, "run-0", *input.ParentRunID)

	require.Len(t, input.Messages, 3)
	assert.Equal(t, RoleUser, input.Messages[0].Role)
	assert.Equal(t, "msg-2", input.Messages[1].ID)
	require.Len(t, input.Messages[1].ToolCalls, 1)
	assert.Equal(t, "tool", input.Messages[1].ToolCalls[0].Function.Name)
	assert.Equal(t, "enc-content-msg-2", input.Messages[1].EncryptedContent)

	assert.Equal(t, RoleReasoning, input.Messages[2].Role)
	assert.Equal(t, "enc-reasoning-1", input.Messages[2].EncryptedValue)
	content, ok := input.Messages[2].ContentString()
	require.True(t, ok)
	assert.Equal(t, "summary", content)

	require.Len(t, input.Tools, 1)
	assert.Equal(t, "tool", input.Tools[0].Name)

	require.Len(t, input.Context, 1)
	assert.Equal(t, "ctx", input.Context[0].Description)

	forwarded, ok := input.ForwardedProps.(map[string]any)
	require.True(t, ok)
	assert.Equal(t, "abc", forwarded["traceId"])
}

// TestRunAgentInputUnmarshalSnakeCase verifies decoding a snake_case RunAgentInput payload.
func TestRunAgentInputUnmarshalSnakeCase(t *testing.T) {
	payload := []byte(`{
		"thread_id": "thread-2",
		"run_id": "run-2",
		"parent_run_id": "run-1",
		"state": {"mode": "snake"},
		"messages": [
			{
				"id": "msg-1",
				"role": "assistant",
				"content": "hi",
				"encrypted_content": "enc-content-msg-1",
				"tool_calls": [
					{
						"id": "tc-2",
						"type": "function",
						"function": {"name": "tool", "arguments": "{\"x\":1}"}
					}
				]
			},
			{
				"id": "msg-2",
				"role": "tool",
				"content": "ok",
				"encrypted_content": "enc-content-msg-2",
				"encrypted_value": "enc-msg-2",
				"tool_call_id": "tc-2",
				"error": "failed"
			},
			{
				"id": "msg-3",
				"role": "activity",
				"activity_type": "progress",
				"content": {"step": 1}
			},
			{
				"id": "reasoning-2",
				"role": "reasoning",
				"content": "thinking",
				"encrypted_value": "enc-reasoning-2"
			}
		],
		"tools": [],
		"context": [],
		"forwarded_props": {"trace_id": "xyz"}
	}`)

	var input RunAgentInput
	err := json.Unmarshal(payload, &input)
	require.NoError(t, err)

	assert.Equal(t, "thread-2", input.ThreadID)
	assert.Equal(t, "run-2", input.RunID)
	require.NotNil(t, input.ParentRunID)
	assert.Equal(t, "run-1", *input.ParentRunID)

	require.Len(t, input.Messages, 4)
	assert.Equal(t, RoleAssistant, input.Messages[0].Role)
	assert.Equal(t, "enc-content-msg-1", input.Messages[0].EncryptedContent)
	require.Len(t, input.Messages[0].ToolCalls, 1)
	assert.Equal(t, "tc-2", input.Messages[0].ToolCalls[0].ID)

	assert.Equal(t, RoleTool, input.Messages[1].Role)
	assert.Equal(t, "tc-2", input.Messages[1].ToolCallID)
	assert.Equal(t, "failed", input.Messages[1].Error)
	assert.Equal(t, "enc-content-msg-2", input.Messages[1].EncryptedContent)
	assert.Equal(t, "enc-msg-2", input.Messages[1].EncryptedValue)
	require.IsType(t, "", input.Messages[1].Content)
	assert.Equal(t, "ok", input.Messages[1].Content.(string))

	assert.Equal(t, RoleActivity, input.Messages[2].Role)
	assert.Equal(t, "progress", input.Messages[2].ActivityType)
	require.IsType(t, map[string]any{}, input.Messages[2].Content)
	contentMap := input.Messages[2].Content.(map[string]any)
	assert.Equal(t, float64(1), contentMap["step"])

	assert.Equal(t, RoleReasoning, input.Messages[3].Role)
	assert.Equal(t, "enc-reasoning-2", input.Messages[3].EncryptedValue)
	content, ok := input.Messages[3].ContentString()
	require.True(t, ok)
	assert.Equal(t, "thinking", content)

	forwarded, ok := input.ForwardedProps.(map[string]any)
	require.True(t, ok)
	assert.Equal(t, "xyz", forwarded["trace_id"])
}

// TestInputContentUnmarshalSnakeCase verifies decoding snake_case fields in InputContent.
func TestInputContentUnmarshalSnakeCase(t *testing.T) {
	payload := []byte(`{
		"type": "binary",
		"mime_type": "image/png",
		"url": "https://example.com/test.png",
		"filename": "test.png"
	}`)

	var content InputContent
	err := json.Unmarshal(payload, &content)
	require.NoError(t, err)

	assert.Equal(t, InputContentTypeBinary, content.Type)
	assert.Equal(t, "image/png", content.MimeType)
	assert.Equal(t, "https://example.com/test.png", content.URL)
	assert.Equal(t, "test.png", content.Filename)
}

// TestInputContentUnmarshalBinaryRequiresSource verifies binary InputContent requires at least one source field.
func TestInputContentUnmarshalBinaryRequiresSource(t *testing.T) {
	payload := []byte(`{
		"type": "binary",
		"mimeType": "image/png"
	}`)

	var content InputContent
	err := json.Unmarshal(payload, &content)
	assert.Error(t, err)
}

// TestInputContentUnmarshalBinaryRequiresMimeType verifies binary InputContent requires a mimeType field.
func TestInputContentUnmarshalBinaryRequiresMimeType(t *testing.T) {
	payload := []byte(`{
		"type": "binary",
		"url": "https://example.com/test.png"
	}`)

	var content InputContent
	err := json.Unmarshal(payload, &content)
	assert.Error(t, err)
}

// TestMessageContentString verifies ContentString extracts text content.
func TestMessageContentString(t *testing.T) {
	msg := Message{Role: RoleAssistant, Content: "hello"}
	text, ok := msg.ContentString()
	assert.True(t, ok)
	assert.Equal(t, "hello", text)

	msg = Message{Role: RoleAssistant, Content: []any{}}
	_, ok = msg.ContentString()
	assert.False(t, ok)

	msg = Message{Role: RoleActivity, Content: "hello"}
	_, ok = msg.ContentString()
	assert.False(t, ok)
}

// TestMessageContentInputContents verifies ContentInputContents extracts multimodal input parts.
func TestMessageContentInputContents(t *testing.T) {
	payload := []byte(`{
		"id": "msg-1",
		"role": "user",
		"content": [
			{"type": "text", "text": "hi"},
			{"type": "binary", "mime_type": "image/png", "url": "https://example.com/test.png"}
		]
	}`)

	var msg Message
	err := json.Unmarshal(payload, &msg)
	require.NoError(t, err)

	parts, ok := msg.ContentInputContents()
	require.True(t, ok)
	require.Len(t, parts, 2)
	assert.Equal(t, InputContentTypeText, parts[0].Type)
	assert.Equal(t, "hi", parts[0].Text)
	assert.Equal(t, InputContentTypeBinary, parts[1].Type)
	assert.Equal(t, "image/png", parts[1].MimeType)
	assert.Equal(t, "https://example.com/test.png", parts[1].URL)

	msg = Message{Role: RoleUser, Content: "plain"}
	_, ok = msg.ContentInputContents()
	assert.False(t, ok)

	msg = Message{
		Role: RoleUser,
		Content: []InputContent{
			{Type: InputContentTypeBinary, MimeType: "image/png"},
		},
	}
	_, ok = msg.ContentInputContents()
	assert.False(t, ok)

	msg = Message{
		Role: RoleUser,
		Content: []InputContent{
			{Type: InputContentTypeBinary, URL: "https://example.com/test.png"},
		},
	}
	_, ok = msg.ContentInputContents()
	assert.False(t, ok)

	msg = Message{
		Role: RoleUser,
		Content: []InputContent{
			{Type: InputContentTypeBinary, MimeType: "image/png", URL: "https://example.com/test.png"},
		},
	}
	parts, ok = msg.ContentInputContents()
	require.True(t, ok)
	require.Len(t, parts, 1)
	assert.Equal(t, "https://example.com/test.png", parts[0].URL)
}

// TestMessageContentActivity verifies ContentActivity extracts structured activity content.
func TestMessageContentActivity(t *testing.T) {
	payload := []byte(`{
		"id": "msg-1",
		"role": "activity",
		"activityType": "progress",
		"content": {"step": 1}
	}`)

	var msg Message
	err := json.Unmarshal(payload, &msg)
	require.NoError(t, err)

	content, ok := msg.ContentActivity()
	require.True(t, ok)
	assert.Equal(t, float64(1), content["step"])

	msg = Message{Role: RoleActivity, Content: "plain"}
	_, ok = msg.ContentActivity()
	assert.False(t, ok)

	msg = Message{Role: RoleUser, Content: map[string]any{"step": 1}}
	_, ok = msg.ContentActivity()
	assert.False(t, ok)
}
